src/WinRT.Runtime/Attributes.cs

```diff
@@ -131,30 +153,80 @@ interface IWinRTExposedTypeDetails
+        /// </remarks>
         public WinRTExposedTypeAttribute()
         {
         }
 
+        /// <summary>
+        /// Creates a new <see cref="WinRTExposedTypeAttribute"/> instance with the specified parameters.
+        /// </summary>
+        /// <param name="winrtExposedTypeDetails">The <see cref="IWinRTExposedTypeDetails"/> implementation type to use to retrieve vtable entries for the annotated type.</param>
         public WinRTExposedTypeAttribute(
             [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
             Type winrtExposedTypeDetails)
         {
             WinRTExposedTypeDetails = winrtExposedTypeDetails;
         }
 
+        /// <inheritdoc cref="IWinRTExposedTypeDetails.GetExposedInterfaces"/>
         public ComWrappers.ComInterfaceEntry[] GetExposedInterfaces()
         {
-            return WinRTExposedTypeDetails != null ? 
+            return WinRTExposedTypeDetails is not null ? 
                 ((IWinRTExposedTypeDetails)Activator.CreateInstance(WinRTExposedTypeDetails)).GetExposedInterfaces() : 
                 Array.Empty<ComWrappers.ComInterfaceEntry>();
         }
 
-#if NET
-        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
-#endif
+#if NET
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
+#endif
         internal Type WinRTExposedTypeDetails { get; }
     }
 
+#if NET8_0_OR_GREATER
+    /// <summary>
+    /// An <see cref="IWinRTExposedTypeDetails"/> implementation for types that are explicitly not meant to be marshalled to native code.
+    /// </summary>
+    /// <remarks>
+    /// This type can be used as argument for <see cref="WinRTExposedTypeAttribute"/> on any managed types that are not meant to be marshalled,
+    /// but that would otherwise trigger the CsWinRT AOT generator (because they implement a WinRT interface, or derive from a WinRT exposed
+    /// type). When support for marshalling is not needed, using this type allows suppressing warnings, and disabling that code generation.
+    /// This can also provide a small binary size improvement in such cases, as that unnecessary marshalling code will not be emitted.
+    /// </remarks>
+#if EMBED
+    internal
+#else
+    public
+#endif
+    sealed class WinRTManagedOnlyTypeDetails : IWinRTExposedTypeDetails
+    {
+        /// <inheritdoc/>
+        public ComWrappers.ComInterfaceEntry[] GetExposedInterfaces()
+        {
+            ThrowNotSupportedException();
+
+            return Array.Empty<ComWrappers.ComInterfaceEntry>();
+        }
+
+        /// <summary>
+        /// Throws a <see cref="NotSupportedException"/> to signal that a blocked type is being marshalled.
+        /// </summary>
+        [DoesNotReturn]
+        private static void ThrowNotSupportedException()

```

Comment:
Can we add a test to the ccw functional test that confirms an exception is thrown.

Reply:
Done in https://github.com/microsoft/CsWinRT/pull/1815/commits/9846e0f7650583b805a3585e77a39be8f98e808c!